### ISSUE TITLE
ONNX export for multilabel - discrepancy between pytorch and onnx outputs

### ISSUE THREAD

**Issue by nishanthcmesh** [huggingface/setfit#291]
Hi,
I was trying out exporting a multilabel setfit model with onnx and the exported model was giving different outputs as compared to the pytorch one. Am I doing something wrong in my inference code? Is multilabel onnx support not there at the moment?

Here is a [Colab](https://colab.research.google.com/drive/1dfR5lh1d0kJ0Sl5BuhtIclcqmJI8FEz9?usp=sharing) to recreate the issue.

**Comment by nishanthcmesh** [comment]
UPDATE: I see that there is a ```torch.sigmoid``` layer added to the outputs for multitarget in ```SetFitHead```. Does this mean I will just have to add the same and check for >0.5 ?

**Comment by tomaarsen** [comment]
Hello!

You're not doing anything wrong I believe, this simply does not work (yet). I ran into this a while ago when adding some more tests for the ONNX exporter, see [here](https://github.com/huggingface/setfit/blob/main/tests/exporters/test_onnx.py#L57). Perhaps @nbertagnolli, who authored the ONNX support for SetFit, can have a look at this?

- Tom Aarsen

**Comment by nbertagnolli** [comment]
Let me take a look! Sorry that it wasn't working for you : (.

**Comment by nbertagnolli** [comment]
I've taken a look and it seems like there are some differences between the predictions when converting mpnet based models to ONNX. I'm not sure this is an issue with the SetFit implementation of the ONNX process.  If you convert just the "sentence-transformers/paraphrase-mpnet-base-v2" model to ONNX using torch.onnx.export there are numerical differences in the output.  To see this try running with a non mpnet model like `sentence-transformers/all-MiniLM-L6-v2`.

Some code below to play with this if you'd like to try for yourself.

`from setfit import SetFitModel
from sentence_transformers import SentenceTransformer
from setfit import SetFitHead, SetFitHead, SetFitModel
from setfit.exporters.onnx import export_onnx
from setfit.exporters.utils import mean_pooling
from transformers import AutoTokenizer
import torch
import onnxruntime

model_id = "sentence-transformers/paraphrase-mpnet-base-v2"
model_id = "sentence-transformers/all-MiniLM-L6-v2"
model_body = SentenceTransformer(model_id)
model_head = SetFitHead(in_features = model_body.get_sentence_embedding_dimension(), out_features = 4, multitarget = True)

model = SetFitModel(model_body = model_body, model_head = model_head, multi_target_strategy="multi-output")
print(model.model_body.training)
print(model.model_head.training)
model.model_body.eval()
model.model_head.eval()
print(model.model_body.training)
print(model.model_head.training)


# Export just the paraphrase-mpnet-base-v2 model
output_path = "setfit_onnx"
max_length = model.model_body.max_seq_length

tokenizer_kwargs = dict(
    max_length=max_length,
    padding="max_length",
    return_attention_mask=True,
    return_token_type_ids=True,
    return_tensors="pt",
)

opset = 12

dummy_sample = "It's a test."
inputs = model.model_body.tokenizer(dummy_sample, **tokenizer_kwargs)

input_names = list(inputs.keys())
output_names = ["logits"]

# Setup the dynamic axes for onnx conversion.
dynamic_axes_input = {}
for input_name in input_names:
    dynamic_axes_input[input_name] = {0: "batch_size", 1: "sequence"}

dynamic_axes_output = {}
for output_name in output_names:
    dynamic_axes_output[output_name] = {0: "batch_size"}

model.model_body.eval()
with torch.no_grad():
    torch.onnx.export(
        model.model_body._modules["0"].auto_model,
        args=tuple(inputs.values()),
        f=output_path,
        opset_version=opset,
        input_names=["input_ids", "attention_mask", "token_type_ids"],
        output_names=output_names,
        dynamic_axes={**dynamic_axes_input, **dynamic_axes_output},
    )
    
sample_text = "whats up"

# Tokenize sentences
encoded_input = model.model_body.tokenizer([sample_text], **tokenizer_kwargs)

# Evaluate on original sentence transformer
model_output = model.model_body._modules["0"].auto_model(**encoded_input)
sentence_embeddings = mean_pooling(model_output[0], encoded_input['attention_mask'])
print(sentence_embeddings)

inputs = model.model_body.tokenizer(
    sample_text,
    max_length=max_length,
    padding="max_length",
    truncation=True,
    return_attention_mask=True,
    return_token_type_ids=True,
    return_tensors="np",
)

session = onnxruntime.InferenceSession(output_path)

onnx_preds = session.run(None, dict(inputs))[0]
print(torch.Tensor(onnx_preds).shape)
print(mean_pooling(torch.Tensor(onnx_preds), encoded_input['attention_mask']))`

**Comment by radandreicristian-cnx** [comment]
Hi @nbertagnolli , I just wanted to confirm that other models (miniLM-L6 and miniLM-L12) are also not working properly after onnx export for multi-label inference.

**Comment by parthjain99** [comment]
miniLM-L6 worked fine for me although Mpnet-base was acting funny. But cant trust it until this thread is closed.

**Comment by kjoth** [comment]
HI @nishanthcmesh , What is the practical application of this approach? In this scenario, there is no fine-tuning of the Sentence Transformer (ST) model. So, what is the rationale behind utilizing SetFit?

**Comment by geraldstanje** [comment]
i currently read about:
```
As of PyTorch 2.1, there are two versions of ONNX Exporter.

torch.onnx.dynamo_export is the newest (still in beta) exporter based on the TorchDynamo technology released with PyTorch 2.0. - looks like its stable now: https://pytorch.org/docs/stable/onnx_dynamo.html

torch.onnx.export is based on TorchScript backend and has been available since PyTorch 1.2.0.
```

can you also use torch.onnx.dynamo_export to export a setfit model?

**Comment by Ulipenitz** [comment]
Hi @nbertagnolli! We are also waiting for this to be resolved, just wanted to ask if it is planned to include it in the next release?

**Comment by nishanthcgit** [comment]
> HI @nishanthcmesh , What is the practical application of this approach? In this scenario, there is no fine-tuning of the Sentence Transformer (ST) model. So, what is the rationale behind utilizing SetFit?

Sorry for replying so late but the issue is that there is some issue in the onnx export which causes the model output to be different than the pytorch version. For verifying this, it doesn't matter whether the model is trained or not.

**Comment by geraldstanje** [comment]
@nishanthcgit did you compare the code with latest version of lib?
